CMakeLists.txt

```diff
@@ -0,0 +1,167 @@
+cmake_minimum_required(VERSION 3.17)
+
+include(FetchContent)
+include(ExternalProject)
+
+# Detect the system architecture
+EXECUTE_PROCESS(
+    COMMAND uname -m
+    COMMAND tr -d '\n'
+    OUTPUT_VARIABLE ARCHITECTURE
+)
+
+if("${ARCHITECTURE}" STREQUAL "x86_64")
+    message("Building JSON for x86_64")
+elseif("${ARCHITECTURE}" STREQUAL "aarch64")
+    message("Building JSON for aarch64")
+else()
+    message(FATAL_ERROR "Unsupported architecture. JSON is only supported on x86_64 and aarch64.")
+endif()
+
+# Project definition
+project(ValkeyJSONModule VERSION 1.0 LANGUAGES C CXX)
+
+# Set the name of the JSON shared library
+set(JSON_MODULE_LIB json)
+
+# Define the Valkey directories
+set(VALKEY_DOWNLOAD_DIR "${CMAKE_BINARY_DIR}/_deps/valkey-src")
+set(VALKEY_BIN_DIR "${CMAKE_BINARY_DIR}/_deps/valkey-src/src/valkey/src")
+
+# Download and build Valkey
+ExternalProject_Add(
+    valkey
+    GIT_REPOSITORY https://github.com/valkey-io/valkey.git  # Replace with actual URL
+    GIT_TAG ${VALKEY_VERSION}
+    PREFIX ${VALKEY_DOWNLOAD_DIR}
+    CONFIGURE_COMMAND ""
+    BUILD_COMMAND make distclean && make -j
+    INSTALL_COMMAND ""
+    BUILD_IN_SOURCE 1
+)
+
+# Define the paths for the copied files
+set(VALKEY_INCLUDE_DIR "${CMAKE_CURRENT_SOURCE_DIR}/src/include")
+set(VALKEY_BINARY_DEST "${CMAKE_CURRENT_SOURCE_DIR}/tst/integration/.build/binaries/${VALKEY_VERSION}")
+
+ExternalProject_Add_Step(
+    valkey
+    copy_header_files
+    COMMENT "Copying header files to include/ directory"
+    DEPENDEES download
+    DEPENDERS configure
+    COMMAND ${CMAKE_COMMAND} -E make_directory ${VALKEY_INCLUDE_DIR}
+    COMMAND ${CMAKE_COMMAND} -E copy ${VALKEY_DOWNLOAD_DIR}/src/valkey/src/valkeymodule.h ${VALKEY_INCLUDE_DIR}/valkeymodule.h
+    ALWAYS 1
+)
+
+# Copy header and binary after Valkey make
+add_custom_command(TARGET valkey
+    POST_BUILD
+    COMMAND ${CMAKE_COMMAND} -E make_directory ${VALKEY_BINARY_DEST}
+    COMMAND ${CMAKE_COMMAND} -E copy ${VALKEY_BIN_DIR}/valkey-server ${VALKEY_BINARY_DEST}/valkey-server
+    COMMENT "Copied valkeymodule.h and valkey-server to destination directories"
+)
+
+# Define valkey-bloom branch
+set(VALKEY_BLOOM_BRANCH "unstable" CACHE STRING "Valkey-bloom branch to use")
+
+# Set the download directory for Valkey-bloom
+set(VALKEY_BLOOM_DOWNLOAD_DIR "${CMAKE_CURRENT_BINARY_DIR}/_deps/valkey-bloom-src")
+
+# Download valkey-bloom
+ExternalProject_Add(
+    valkey-bloom
+    GIT_REPOSITORY https://github.com/valkey-io/valkey-bloom.git
```

Comment:
I would like to see some way to not download dependencies during the build, at least optionally.

Offline builds are common to guarantee that a project doesn't have any online dependencies that have not been verified in advance. Supply chain attacks is the concern.

I suggest we add some environmant variable to provide the dependencies using a local paths and CMake can use these instead of downloading them if they are provided. It can be added in the future though.

Reply:
> I would like to see some way to not download dependencies during the build, at least optionally.
> 
> Offline builds are common to guarantee that a project doesn't have any online dependencies that have not been verified in advance. Supply chain attacks is the main concern. (License issues is another concern.)
> 
> I suggest we add some environmant variable to provide the dependencies using a local paths and CMake can use these instead of downloading them if they are provided. It can be added in the future though.

Reply:
Yes, it's all about reusing the pytest infrastructure in valkey-bloom.

Reply:
https://github.com/valkey-io/valkeyJSON/issues/10